rule/add_constant.go

```diff
@@ -41,8 +42,13 @@ type AddConstantRule struct {
 }
 
 // Apply applies the rule to given file.
-func (r *AddConstantRule) Apply(file *lint.File, arguments lint.Arguments) []lint.Failure {
-	r.configureOnce.Do(func() { r.configure(arguments) })
+func (r *AddConstantRule) Apply(file *lint.File, arguments lint.Arguments) ([]lint.Failure, error) {
+	var configureErr error
+	r.configureOnce.Do(func() { configureErr = r.configure(arguments) })
+
+	if configureErr != nil {
+		return nil, configureErr
+	}
```

Comment:
I Iooked at it again, and there is something I don't get.

The code was

```go
        r.configureOnce.Do(func() { r.configure(arguments) })
```

When there were no problems, r.configure is called only once, even when Add is called multiple times.

Also, r.configure was raising panic
The panic was stopping everything.



Now, the behavior is 

```go
        var configureErr error 
         r.configureOnce.Do(func() { configureErr = r.configure(arguments) }) 
  
         if configureErr != nil { 
                 return nil, configureErr 
         }
```

We kept the behavior of having r.configure being called only once when there is no problem.

But, the behavior differs when an error occurs.

configureErr will be filled with the error only on the first call to Add. The same way the code was doing with a panic, somehow.

So consecutive calls to add won't return an error.

So we are relying on the fact the code calling Add, will stop everything.

It might be OK, but I would like to share my understanding.


So here, either:
- we add a comment that acknowledge,
 it's done on purpose
- we think about a better solution

Reply:
Please note, I'm not asking author to find a solution, but really we find one together

Reply:
The usual solution to this problem is memoization: store the result of the call to `configure` (error or nil) and initialize `configureError`  with that value. 

```go
configureError:= r.configureError
r.configureOnce.Do(func() { configureErr = r.configure(arguments) })
```

Reply:
Yes, I also thought about it, as a possible solution, but it adds a another variable to the struct, but why not after all

Reply:
resolved with ``` return []lint.Failure{lint.NewInternalFailure(configureErr.Error())} ```